[Site Isolation] Disowned opener is not persisted after cross-site navigation

szewai opened the pull request:
#### 2e9bc3ae2ab812cb39de01af0534565aa706915d
<pre>
[Site Isolation] Disowned opener is not persisted after cross-site navigation
<a href="https://rdar.apple.com/165741304">rdar://165741304</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303452">https://bugs.webkit.org/show_bug.cgi?id=303452</a>

Reviewed by Alex Christensen.

Opened window can disown its opener by setting window.opener to null in JavaScript, and in this case `window.opener`
should stay null across navigation. Currently the state is not persisted under site isolation as the web process that
sets opener to null does not inform UI processes about the update. So when cross-site navigation happens and the frame
is loaded in a different process, UI process will tell the new process that the frame has opener. To fix it, make web
process notify UI process when a frame disowns opener on JavaScript request.

This patch also introduces `WebFrameProxy::m_disownedOpener` to track opener that has been disowned, because this change
causes process no longer to be swapped when opener window still has a handle to opened window, when site isolation is
disabled. This cause is the current implementation just relies on `WebFrameProxy::opener()` to track the opener
relationship (see `WebProcessPool::processForNavigationInternal()`). To keep this case still works, this patch uses
`WebFrameProxy::disownedOpener()` in the condition.

Test: SiteIsolation.NavigateOpenedWindowCrossSiteAfterDisowningOpener

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::updateOpener):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::updateOpener):
(WebCore::Frame::disownOpener):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::disownOpener):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::updateOpener):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::opener const):
(WebKit::WebFrameProxy::disownedOpener const):
(WebKit::WebFrameProxy::opener): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateOpener):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateOpener):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
(WebFrameLoaderClient::updateOpener):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateOpenedWindowCrossSiteAfterDisowningOpener)):

Canonical link: <a href="https://commits.webkit.org/304171@main">https://commits.webkit.org/304171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0e6f4b8e05ce836f4c886f28059c063995abc94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134632 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142157 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86580 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7931660d-81c3-4e88-979f-2d14391876c7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136502 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6945 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102905 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70193 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b08559f0-4f04-4d2c-a828-9e3c8b23eda8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137579 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5396 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83708 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8a27fff7-0398-427d-96f8-e7eb6486a0da) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5253 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2864 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2750 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114478 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144851 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6764 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39395 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111299 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6838 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111590 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28336 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5089 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116947 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60652 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6815 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35140 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6624 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70398 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6860 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6733 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2291,7 +2291,7 @@ std::tuple<Ref<WebProcessProxy>, RefPtr<SuspendedPageProxy>, ASCIILiteral> WebPr
     bool isRequestFromClientOrUserInput = navigation.isRequestFromClientOrUserInput() && !navigation.substituteData();
     if (navigation.openedByDOMWithOpener()
-        && page.mainFrame() && page.mainFrame()->opener()
+        && page.mainFrame() && (page.mainFrame()->opener() || page.mainFrame()->disownedOpener())
```
It seems like this might need to be `page.mainFrame()->opener() && ! page.mainFrame()->disownedOpener()`

It also seems like we shouldn't remember the disowned opener at all.  Can we do that instead?

szewai replied:
I don't quite understand this comment.
Before this PR, `WebFrameProxy::opener()` is not null after `window.opener = null`, so this condition is true and process will be swapped under PSON (note this condition check is skipped for site isolation).
After this PR, `WebFrameProxy::opener()` is null after `window.opener = null`, so this condition is skipped and process is not swapped.
In order to keep existing behavior, I added `WebFrameProxy::disownedOpener()` to record the opener that has been disowned, so we can keep this condition working.

szewai replied:
Update the PR as discussed offline.

szewai committed:
[Site Isolation] Disowned opener is not persisted after cross-site navigation
rdar://165741304
https://bugs.webkit.org/show_bug.cgi?id=303452

Reviewed by Alex Christensen.

Opened window can disown its opener by setting window.opener to null in JavaScript, and in this case `window.opener`
should stay null across navigation. Currently the state is not persisted under site isolation as the web process that
sets opener to null does not inform UI processes about the update. So when cross-site navigation happens and the frame
is loaded in a different process, UI process will tell the new process that the frame has opener. To fix it, make web
process notify UI process when a frame disowns opener on JavaScript request.

This patch also introduces `WebFrameProxy::m_disownedOpener` to track opener that has been disowned, because this change
causes process no longer to be swapped when opener window still has a handle to opened window, when site isolation is
disabled. This cause is the current implementation just relies on `WebFrameProxy::opener()` to track the opener
relationship (see `WebProcessPool::processForNavigationInternal()`). To keep this case still works, this patch uses
`WebFrameProxy::disownedOpener()` in the condition.

Test: SiteIsolation.NavigateOpenedWindowCrossSiteAfterDisowningOpener

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::updateOpener):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::updateOpener):
(WebCore::Frame::disownOpener):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::disownOpener):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::updateOpener):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::opener const):
(WebKit::WebFrameProxy::disownedOpener const):
(WebKit::WebFrameProxy::opener): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateOpener):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::updateOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateOpener):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
(WebFrameLoaderClient::updateOpener):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateOpenedWindowCrossSiteAfterDisowningOpener)):

Canonical link: https://commits.webkit.org/304171@main
